src/include/graph/graph.h

```diff
@@ -45,6 +36,68 @@ struct GraphScanResult {
  */
 class Graph {
 public:
+    // Virtual Implementation of iterator functions
+    // Iterators can't really be virtual, but we can create an iterator (GraphResult::iterator)
+    // which uses a polymorphic object for its implementation
+    class IteratorImpl {
+    public:
+        virtual ~IteratorImpl() = default;
+    };
+
+    class Iterator {
+    public:
+        explicit Iterator(GraphScanState* scanState) : scanState{scanState} {}
+        DEFAULT_BOTH_MOVE(Iterator);
+        Iterator(const Iterator& other) = default;
+
+        using iterator_category = std::input_iterator_tag;
+        using difference_type = std::ptrdiff_t;
+        using value_type =
+            std::pair<std::span<const common::nodeID_t>, std::span<const common::relID_t>>;
+        using pointer = value_type;
+        using reference = value_type;
+
+        std::pair<std::span<const common::nodeID_t>, std::span<const common::nodeID_t>>
+        operator*() const {
+            return std::make_pair(scanState->getNodes(), scanState->getEdges());
+        }
+        Iterator& operator++() {
+            if (!scanState->next()) {
+                scanState = nullptr;
+            }
+            return *this;
+        }
+        void operator++(int) { ++*this; }
+        bool operator==(const Iterator& other) const {
+            // Only needed for comparing to the end, so they are equal if and only if both are null
+            return scanState == nullptr && other.scanState == nullptr;
+        }
+        // Counts and consumes the iterator
+        uint64_t count() {
+            // TODO(bmwinger): avoid scanning if all that's necessary is to count the results
+            uint64_t result = 0;
+            do {
+                result += scanState->getNodes().size();
+            } while (scanState->next());
+            return result;
+        }
+
+        std::vector<common::nodeID_t> collectNodes() {
+            std::vector<common::nodeID_t> adjacentNodes;
```

Comment:
Would it better to rename `adjacentNodes` to `nbrNodes`, though they mean the same? I think we should be more constrained to use multiple different terms referring to the same thing.

Reply:
Oh, so that's an abbreviation of neighbour? I thought it was number and was a little confused when I'd seen it elsewhere.

Reply:
😄 yeah we use that for abbr of neighbour.